File: scripts/test_image_library.py
from playwright.sync_api import sync_playwright
import time
import json

SCREENSHOTS_DIR = "C:/Users/User/Prompt-Generator/screenshots"

def capture(page, name):
    path = f"{SCREENSHOTS_DIR}/{name}.png"
    page.screenshot(path=path, full_page=False)
    print(f"Saved: {path}")
    return path

def run():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={'width': 1920, 'height': 1080})
        page = context.new_page()

        # Step 1: Go to http://localhost:3001
        print("Step 1: Navigating to http://localhost:3001 ...")
        page.goto("http://localhost:3001", wait_until='networkidle')
        capture(page, "step1_initial_page")
        print("Step 1 done.")

        # Step 2: Inject 10 real test images via localStorage
        print("Step 2: Injecting test images into localStorage ...")
        inject_result = page.evaluate("""
            () => {
                const images = Array.from({length:10}, (_, i) => ({
                    id: `img-test-${i}-${Date.now()}`,
                    created_at: new Date(Date.now() - i * 60000).toISOString(),
                    filename: `chatgpt-image-${i}.png`,
                    provider: i % 3 === 0 ? 'gemini' : i % 3 === 1 ? 'chatgpt' : 'edit',
                    aspect_ratio: '16:9',
                    resolution: '1K',
                    storage_path: '',
                    public_url: `https://picsum.photos/seed/${i+50}/800/450`
                }));
                localStorage.setItem('pg_generated_images', JSON.stringify(images));
                const stored = JSON.parse(localStorage.getItem('pg_generated_images'));
                console.log('Stored images:', stored.length);
                return { count: stored.length, images: stored };
            }
        """)
        print(f"Injected {inject_result['count']} images into localStorage.")
        print("Image IDs:", [img['id'] for img in inject_result['images']])

        # Step 3: Screenshot console output confirmation
        capture(page, "step3_after_injection")
        print("Step 3: Screenshot after injection taken.")

        # Step 4: Click the "Image Library" button WITHOUT refreshing
        print("Step 4: Looking for 'Image Library' button ...")
        # Try various selectors for the Image Library button
        lib_button = None
        selectors = [
            "text=Image Library",
            "[aria-label='Image Library']",
            "button:has-text('Image Library')",
            "a:has-text('Image Library')",
            "[data-testid='image-library']",
        ]
        for sel in selectors:
            try:
                el = page.locator(sel).first
                if el.is_visible(timeout=2000):
                    lib_button = el
                    print(f"  Found button with selector: {sel}")
                    break
            except Exception:
                continue

        if lib_button is None:
            print("  Could not find 'Image Library' button with standard selectors. Capturing page HTML for inspection.")
            html_snippet = page.evaluate("() => document.body.innerHTML.substring(0, 3000)")
            print("HTML snippet:", html_snippet[:1000])
            capture(page, "step4_no_button_found")
        else:
            lib_button.click()
            print("Step 4: Clicked 'Image Library' button.")

            # Step 5: Screenshot IMMEDIATELY (< 0.3s) — check for skeleton boxes
            time.sleep(0.2)
            capture(page, "step5_immediate_after_click")
            print("Step 5: Immediate screenshot taken (check for skeleton boxes).")

            # Step 6: Wait 4 seconds, take another screenshot — did images load?
            time.sleep(4)
            capture(page, "step6_after_4_seconds")
            print("Step 6: Screenshot after 4 seconds taken (check if images loaded).")

            # Step 7: Click the "Back" button to return to main page
            print("Step 7: Looking for 'Back' button ...")
            back_button = None
            back_selectors = [
                "text=Back",
                "button:has-text('Back')",
                "[aria-label='Back']",
                "[aria-label='back']",
                "a:has-text('Back')",
            ]
            for sel in back_selectors:
                try:
                    el = page.locator(sel).first
                    if el.is_visible(timeout=2000):
                        back_button = el
                        print(f"  Found Back button with selector: {sel}")
                        break
                except Exception:
                    continue

            if back_button is None:
                print("  Could not find 'Back' button. Capturing page for inspection.")
                capture(page, "step7_no_back_button")
            else:
                back_button.click()
                time.sleep(1)
                capture(page, "step7_after_back")
                print("Step 7: Clicked Back, screenshot taken.")

                # Step 8: Click "Image Library" again (second visit)
                print("Step 8: Clicking 'Image Library' again ...")
                lib_button2 = None
                for sel in selectors:
                    try:
                        el = page.locator(sel).first
                        if el.is_visible(timeout=2000):
                            lib_button2 = el
                            print(f"  Found button with selector: {sel}")
                            break
                    except Exception:
                        continue

                if lib_button2 is None:
                    print("  Could not find 'Image Library' button on second visit.")
                    capture(page, "step8_no_button")
                else:
                    lib_button2.click()
                    print("Step 8: Clicked 'Image Library' (second visit).")

                    # Step 9: Screenshot IMMEDIATELY — do images appear WITHOUT skeletons?
                    time.sleep(0.2)
                    capture(page, "step9_second_visit_immediate")
                    print("Step 9: Immediate screenshot on second visit taken.")

                    # Also wait a moment and take a final screenshot
                    time.sleep(3)
                    capture(page, "step9b_second_visit_after_3s")
                    print("Step 9b: Second visit screenshot after 3s taken.")

        browser.close()
        print("\nAll steps complete.")

if __name__ == "__main__":
    run()

File: scripts/test_image_library.mjs
import { chromium } from 'playwright';

const SCREENSHOTS_DIR = 'C:/Users/User/Prompt-Generator/screenshots';

async function capture(page, name) {
  const path = `${SCREENSHOTS_DIR}/${name}.png`;
  await page.screenshot({ path, fullPage: false });
  console.log(`Saved: ${path}`);
  return path;
}

async function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function findButton(page, selectors, timeoutMs = 3000) {
  for (const sel of selectors) {
    try {
      const el = page.locator(sel).first();
      await el.waitFor({ state: 'visible', timeout: timeoutMs });
      console.log(`  Found with selector: ${sel}`);
      return el;
    } catch {
      // try next
    }
  }
  return null;
}

async function run() {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({ viewport: { width: 1920, height: 1080 } });
  const page = await context.newPage();

  // Capture console messages from the page
  page.on('console', msg => console.log(`  [PAGE CONSOLE] ${msg.type()}: ${msg.text()}`));

  // ---------- Step 1: Navigate ----------
  console.log('\n=== Step 1: Navigating to http://localhost:3001 ===');
  await page.goto('http://localhost:3001', { waitUntil: 'networkidle', timeout: 30000 });
  await capture(page, 'step1_initial_page');

  // ---------- Step 2: Inject images into localStorage ----------
  console.log('\n=== Step 2: Injecting 10 test images into localStorage ===');
  const injectResult = await page.evaluate(() => {
    const images = Array.from({ length: 10 }, (_, i) => ({
      id: `img-test-${i}-${Date.now()}`,
      created_at: new Date(Date.now() - i * 60000).toISOString(),
      filename: `chatgpt-image-${i}.png`,
      provider: i % 3 === 0 ? 'gemini' : i % 3 === 1 ? 'chatgpt' : 'edit',
      aspect_ratio: '16:9',
      resolution: '1K',
      storage_path: '',
      public_url: `https://picsum.photos/seed/${i + 50}/800/450`
    }));
    localStorage.setItem('pg_generated_images', JSON.stringify(images));
    const stored = JSON.parse(localStorage.getItem('pg_generated_images'));
    return { count: stored.length, first_id: stored[0].id, last_id: stored[stored.length - 1].id };
  });
  console.log(`  Stored ${injectResult.count} images. First ID: ${injectResult.first_id}, Last ID: ${injectResult.last_id}`);

  // ---------- Step 3: Screenshot after injection ----------
  console.log('\n=== Step 3: Screenshot confirming localStorage injection ===');
  await capture(page, 'step3_after_injection');

  // ---------- Step 4: Click Image Library (NO page refresh) ----------
  console.log('\n=== Step 4: Clicking "Image Library" button (no refresh) ===');
  const libSelectors = [
    'button:has-text("Image Library")',
    'text=Image Library',
    'a:has-text("Image Library")',
    '[aria-label="Image Library"]',
    '[title="Image Library"]',
  ];
  const libButton = await findButton(page, libSelectors);

  if (!libButton) {
    console.log('  ERROR: Could not find Image Library button. Capturing page for inspection.');
    // Dump all button texts
    const buttons = await page.evaluate(() =>
      Array.from(document.querySelectorAll('button, a')).map(el => el.innerText?.trim() || el.getAttribute('aria-label') || el.id).filter(Boolean)
    );
    console.log('  All clickable elements:', buttons.join(' | '));
    await capture(page, 'step4_no_button_found');
    await browser.close();
    return;
  }

  await libButton.click();

  // ---------- Step 5: Immediate screenshot (< 0.3s) — skeleton boxes? ----------
  console.log('\n=== Step 5: Immediate screenshot (~0.2s after click) — checking for skeletons ===');
  await sleep(200);
  await capture(page, 'step5_immediate_0_2s');

  // ---------- Step 6: Wait 4s, screenshot — images loaded? ----------
  console.log('\n=== Step 6: Waiting 4 seconds then screenshot ===');
  await sleep(4000);
  await capture(page, 'step6_after_4_seconds');

  // Check what's visible in the library
  const libraryState = await page.evaluate(() => {
    const imgs = document.querySelectorAll('img');
    const skeletons = document.querySelectorAll('[class*="skeleton"], [class*="Skeleton"], [class*="loading"], [class*="shimmer"], [class*="animate-pulse"]');
    return {
      image_count: imgs.length,
      skeleton_count: skeletons.length,
      img_srcs: Array.from(imgs).slice(0, 5).map(i => i.src),
    };
  });
  console.log(`  Images on page: ${libraryState.image_count}`);
  console.log(`  Skeleton elements: ${libraryState.skeleton_count}`);
  console.log(`  First few img srcs: ${libraryState.img_srcs.join(', ')}`);

  // ---------- Step 7: Click Back ----------
  console.log('\n=== Step 7: Clicking "Back" button ===');
  const backSelectors = [
    'button:has-text("Back")',
    'text=Back',
    'a:has-text("Back")',
    '[aria-label="Back"]',
    '[aria-label="back"]',
    'button:has-text("← Back")',
    'button:has-text("‹")',
  ];
  const backButton = await findButton(page, backSelectors);

  if (!backButton) {
    console.log('  ERROR: Could not find Back button. Dumping clickables...');
    const buttons = await page.evaluate(() =>
      Array.from(document.querySelectorAll('button, a')).map(el => el.innerText?.trim() || el.getAttribute('aria-label')).filter(Boolean)
    );
    console.log('  Clickable elements in library view:', buttons.join(' | '));
    await capture(page, 'step7_no_back_button');
  } else {
    await backButton.click();
    await sleep(800);
    await capture(page, 'step7_after_back');
    console.log('  Back clicked, screenshot taken.');

    // ---------- Step 8: Click Image Library again ----------
    console.log('\n=== Step 8: Second visit — clicking Image Library again ===');
    const libButton2 = await findButton(page, libSelectors);

    if (!libButton2) {
      console.log('  ERROR: Could not find Image Library button on second visit.');
      await capture(page, 'step8_no_button');
    } else {
      await libButton2.click();

      // ---------- Step 9: Immediate screenshot on second visit ----------
      console.log('\n=== Step 9: Immediate screenshot on second visit (~0.2s) ===');
      await sleep(200);
      await capture(page, 'step9_second_visit_immediate');

      const secondVisitState = await page.evaluate(() => {
        const imgs = document.querySelectorAll('img');
        const skeletons = document.querySelectorAll('[class*="skeleton"], [class*="Skeleton"], [class*="loading"], [class*="shimmer"], [class*="animate-pulse"]');
        return {
          image_count: imgs.length,
          skeleton_count: skeletons.length,
        };
      });
      console.log(`  Second visit — Images: ${secondVisitState.image_count}, Skeletons: ${secondVisitState.skeleton_count}`);

      await sleep(3000);
      await capture(page, 'step9b_second_visit_3s');
      console.log('  Second visit 3s screenshot taken.');
    }
  }

  await browser.close();
  console.log('\n=== All steps complete ===');
}

run().catch(err => {
  console.error('Script error:', err);
  process.exit(1);
});
